thriftclient/client.go

```diff
@@ -0,0 +1,67 @@
+package thriftclient
+
+import (
+	"context"
+
+	"github.com/apache/thrift/lib/go/thrift"
+
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// MonitoredClient implements the thrift.TClient interface and can be used to
+// wrap thrift calls in a client span.
+//
+// Calls will only be wrapped in client spans if the given context has a parent
+// span (either a server or "active" local/client span) set on it.  If no such
+// span is set, then the call will still proceed, it will just not be monitored
+// by a client span.
+type MonitoredClient struct {
+	client thrift.TClient
+}
+
+// NewMonitoredClientFromFactory returns a pointer to a new MonitoredClient that
+// a new TStandardClient using the provided transport and protocol factory.
+func NewMonitoredClientFromFactory(transport thrift.TTransport, factory thrift.TProtocolFactory) *MonitoredClient {
+	return NewMonitoredClient(thrift.NewTStandardClient(
+		factory.GetProtocol(transport),
+		factory.GetProtocol(transport),
+	))
+}
+
+// NewMonitoredClient returns a pointer to a new MonitoredClient that wraps the
+// given thrift.TClient calls in thrift client spans.
+func NewMonitoredClient(c thrift.TClient) *MonitoredClient {
+	if c == nil {
+		panic("NewMonitoredClient received a nil client")
+	}
+	return &MonitoredClient{client: c}
+}
+
+// Call wraps the underlying thrift.TClient.Call method with a client span.
+func (c MonitoredClient) Call(ctx context.Context, method string, args, result thrift.TStruct) error {
+	var span *tracing.Span
+	if span = tracing.GetActiveSpan(ctx); span == nil {
+		span = tracing.GetServerSpan(ctx)
+	}
+	var child *tracing.Span
+	var err error
+	if span != nil {
+		ctx, child = span.ChildAndThriftContext(ctx, method)
+	}
+	defer func() {
+		if child != nil {
+			if err = child.Stop(ctx, err); err != nil {
```

Comment:
If you want to access the returning `err` from within the deferred function, the best practice is to use named return (you don't even need to do `err = c.client.Call(...); return err`, just `return c.client.Call(...)` would do). But in that case inside the deferred function you need to be careful not to overwrite it, which will also overwrite the returned err (unless you intend to do so), so here you need to use a different name for the error returned by `child.Stop`.